Web inspector fails to syntax highlight JS with backtick-delimited format strings

rcaliman-apple opened the pull request:
#### fe5c7a50d8e3011d908ed5014acaa9e941225e05
<pre>
Web inspector fails to syntax highlight JS with backtick-delimited format strings
<a href="https://bugs.webkit.org/show_bug.cgi?id=283870">https://bugs.webkit.org/show_bug.cgi?id=283870</a>
<a href="https://rdar.apple.com/107619553">rdar://107619553</a>

Reviewed by Devin Rousso and BJ Burg.

Update CodeMirror from version 5.21.1 to 5.65.20

Highlights
- Syntax highlighting for template literals (template strings)
- Styntax highlighting for private class elements

* Source/WebInspectorUI/Scripts/update-codemirror-resources.rb:
Check for the latest repository name format.
Stop copying over addon/mode/overlay.js and mode/sql/sql.js because they&apos;re not used anymore.

* Source/WebInspectorUI/Tools/CodeMirrorModes/index.html:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/LICENSE:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/clike.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/clojure.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/closebrackets.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/codemirror.css:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/codemirror.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/coffeescript.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/comment.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/css.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/htmlmixed.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/javascript.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/jsx.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/livescript.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/mark-selection.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/matchbrackets.js:

* Source/WebInspectorUI/UserInterface/External/CodeMirror/overlay.js: Removed.
`CodeMirror.overlayMode` Doesn&apos;t seem to be used anywhere.
This file has existed since Web Inspector inception and was periodically updated,
but whatever used this functionality was probably removed at some point.

* Source/WebInspectorUI/UserInterface/External/CodeMirror/placeholder.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/runmode.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/sass.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/searchcursor.js:

* Source/WebInspectorUI/UserInterface/External/CodeMirror/sql.js: Removed.
SQL syntax highlighting is no longer used.

* Source/WebInspectorUI/UserInterface/External/CodeMirror/sublime.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/xml.js:
* Source/WebInspectorUI/UserInterface/Main.html:

* Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:
(WI.CSSStyleDeclaration.prototype.resolveVariableValue):
* Source/WebInspectorUI/UserInterface/Views/CodeMirrorAdditions.js:
* Source/WebInspectorUI/UserInterface/Views/CodeMirrorOverrides.css:
(.CodeMirror pre:is(.CodeMirror-line, .CodeMirror-line-like)):
(.CodeMirror pre): Deleted.
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetStyleProperty.js:
(WI.SpreadsheetStyleProperty.prototype._addGradientTokens):
(WI.SpreadsheetStyleProperty.prototype._addColorTokens):
(WI.SpreadsheetStyleProperty.prototype._addTimingFunctionTokens):
(WI.SpreadsheetStyleProperty.prototype._addVariableTokens):

In CodeMirror version 5.43.0, the token type for CSS functions was changed
from &quot;atom&quot; to &quot;variable callee&quot;: <a href="https://github.com/codemirror/codemirror5/commit/3c79c54a335411fd1aff55246b77131dd98bb81b">https://github.com/codemirror/codemirror5/commit/3c79c54a335411fd1aff55246b77131dd98bb81b</a>
Update references in Web Inspector which check against that token type.

* Source/WebInspectorUI/UserInterface/Views/SyntaxHighlightingDefaultTheme.css:
(.cm-s-default :is(.cm-tag, .cm-bracket, .cm-atom, .cm-variable.cm-callee, .cm-keyword, .cm-m-javascript.cm-builtin),):
(.cm-s-default :is(.cm-tag, .cm-bracket, .cm-atom, .cm-keyword, .cm-m-javascript.cm-builtin),): Deleted.

Update syntax highlighting theme to account for the new `cm-callee` class name set on CSS functions.

* Source/WebInspectorUI/UserInterface/Views/TextEditor.js:
(WI.TextEditor.prototype.toggleStyleClassForLine): Deleted.
toggleStyleClassForLine() is no longer used.

Canonical link: <a href="https://commits.webkit.org/300332@main">https://commits.webkit.org/300332@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/080d5e93a8c2fbfb03f808b5339cabdb6e8e8eca

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122130 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41833 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32502 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128702 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74224 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124006 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42547 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50426 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92825 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61709 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125082 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33935 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109367 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73479 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32948 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27533 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-font-loading/idlharness.https.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72190 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103448 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27724 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131454 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49069 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37326 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101391 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49443 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105581 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101260 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25677 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46630 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24749 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/45819 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48926 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54660 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48396 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51746 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50076 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/External/CodeMirror/sql.js:
@burg After the removal of support for WebSQL in https://commits.webkit.org/289854@main, does it still make sense to keep the SQL CodeMirror mode? 

There's nothing else in Web Inspector that *explicitly* sets that mode. Is there a use case where it's still needed?

rcaliman-apple replied:
The style checker is upset about this file for some of the language used. 
If we don't need it anymore, removing it pacifies that script.

burg replied:
I am pretty sure you can get rid of it

dcrousso replied:
yeah we should be fine removing this

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:
```diff
@@ -393,7 +393,7 @@ WI.CSSStyleDeclaration = class CSSStyleDeclaration extends WI.Object
             for (let i = 0; i < tokens.length; i++) {
                 let token = tokens[i];
-                if (token.value === "var" && token.type && token.type.includes("atom")) {
+                if (token.value === "var" && token.type && token.type.includes("variable callee")) {
```
are there any other things we need to adjust like this?  i think you can search for `token.type` to find most of the things like this

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/CodeMirrorAdditions.js:
```diff
@@ -266,12 +266,11 @@
-                    style = style + " hex-color";
+            if (style === "atom" && hexColorRegex.test(stream.current())) {
+                style = style + " hex-color";
+            } else if (style.includes("variable callee") && stream.current() === "url") {
```
are we sure it's always in this order?

i think it'd be better to check `style.includes("variable") && style.includes("callee")` instead (and frankly it'd be even more ideal to do it as `/\bvariable\b/.test(style) && /\bcallee\b/.test(style)` just in case "variable" or "callee" are a substring of something else)

ditto in the other files too

rcaliman-apple replied:
>are we sure it's always in this order?

Yes, it is always in this order. The token type is hardcoded as `"variable callee"` in `mode/css.js` by the CodeMirror patch that replaced `"atom"` (`"word"` actually, but that gets overridden with `"atom"` later in the file): https://github.com/codemirror/codemirror5/commit/3c79c54a335411fd1aff55246b77131dd98bb81b

>(and frankly it'd be even more ideal to do it as /\bvariable\b/.test(style) && /\bcallee\b/.test(style) just in case "variable" or "callee" are a substring of something else)

That's a good idea. I'll update the checks.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SyntaxHighlightingDefaultTheme.css:
```diff
@@ -38,7 +38,7 @@
 }
 
-.cm-s-default :is(.cm-tag, .cm-bracket, .cm-atom, .cm-keyword, .cm-m-javascript.cm-builtin),
+.cm-s-default :is(.cm-tag, .cm-bracket, .cm-atom, .cm-callee, .cm-keyword, .cm-m-javascript.cm-builtin),
```
is `.cm-atom` still needed?

should this be `.cm-variable.cm-callee` instead?

rcaliman-apple replied:
>is .cm-atom still needed?

Yes, at least color keywords use it. Perhaps others.

>should this be .cm-variable.cm-callee instead?

That also works. Didn't see a need to combine the two to guard against any specificity fight

dcrousso commented on Source/WebInspectorUI/Scripts/update-codemirror-resources.rb:
```diff
@@ -34,7 +34,6 @@ def verify_code_mirror_repository_path(path)
   addon/display/placeholder.js
   addon/edit/closebrackets.js
   addon/edit/matchbrackets.js
-  addon/mode/overlay.js
```
are we sure it's safe to remove this?  there's at least one use of `addOverlay`

rcaliman-apple replied:
The single use of `cm.addOverlay` in `CodeMirrorAdditions.js` seems unrelated to `CodeMirror.overlayMode`
 https://github.com/WebKit/WebKit/blob/e6bb7d08e4b7821cc4f018d0c17e3f918e577868/Source/WebInspectorUI/UserInterface/Views/CodeMirrorAdditions.js#L381
 
 From CodeMirror's docs:
 
 >cm.addOverlay(mode: string|object, ?options: object)
Enable a highlighting overlay. This is a stateless mini-mode that can be used to add extra highlighting. For example, the [search addon](https://codemirror.net/5/demo/search.html) uses it to highlight the term that's currently being searched.

meanwhile...

 >[mode/overlay.js](https://codemirror.net/5/addon/mode/overlay.js)
Mode combinator that can be used to extend a mode with an 'overlay' — a secondary mode is run over the stream, along with the base mode, and can color specific pieces of text without interfering with the base mode. Defines CodeMirror.overlayMode, which is used to create such a mode. See [this demo](https://codemirror.net/5/demo/mustache.html) for a detailed example.

I found no calls to `overlayMode()` anywhere in Web Inspector.

dcrousso reviewed (commented):
in addition to the stylistic changes, please verify that all of the following still work as expected:
- subclasses of `WI.CodeMirrorEditingController` and `WI.Annotator`
- editors using custom MIME types `text/x-regex` and `text/x-local-override-url`
- overridden options and keyboard shortcuts from `WI.CodeMirrorEditor` and `WI.CodeMirrorTextKillController`
- all of the custom logic in `Source/WebInspectorUI/UserInterface/Views/CodeMirrorAdditions.js`
- general behaviors (e.g. jumping to a line, autocomplete, responsive sizing inside a `WI.Popover` while typing, etc.)
- etc.

there are a **_LOT_** of things listed in the [CodeMirror release notes](https://codemirror.net/5/doc/releases.html) so i think we should be extra careful and thorough before updating

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/CodeMirrorAdditions.js:
```diff
@@ -266,12 +266,11 @@
-                    state._expectLink = true;
-                } else if (hexColorRegex.test(stream.current()))
-                    style = style + " hex-color";
+            if (style === "atom" && hexColorRegex.test(stream.current())) {
```
NIT: can we make this `/\batom\b/.test(style)` to match the others?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/CodeMirrorOverrides.css:
```diff
@@ -37,7 +37,8 @@
 
-.CodeMirror pre {
+.CodeMirror pre.CodeMirror-line,
+.CodeMirror pre.CodeMirror-line-like {
```
we can combine these
```css
.CodeMirror pre:is(.CodeMirror-line, .CodeMirror-line-like) {
```

dcrousso reviewed (approved):
nice work!

> I did find a couple of unused files:
> 
> - `WI.CodeMirrorDragToAdjustNumberController`
> - `WI.DragToAdjustController`
> 
> They seem to have been used in an earlier version of a gradient editor. But they don't seem to be used anymore.
Should I remove them in a separate patch?

i think these are still used by our on-by-default `dragToAdjustNumbers` custom option

> Checked editors for regex,  local overrides (the [implementation](https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Views/CodeMirrorLocalOverrideURLMode.js) just seems to highlight invalid scheme and fragment) and some interaction in popovers. 

it's not allowed to create a Local Override for anything other than `http://`/`https://`/`file://` and it cannot have a fragment (though i think we maybe should support that at least for `WI.LocalResourceOverride.InterceptType.ResponseSkippingNetwork` since it never even hits the network? 🤔)

> Also found this [neat tool to test CodeMirror modes](https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/Tools/CodeMirrorModes/index.html) in the repo.

oh yeah i forgot that exists.  please also update that for removing `sql` support

> Found that `CodeMirror.toggleLineClass()` isn't really used anymore. `WI.TextEditor.prototype.toggleStyleClassForLine()` is the only caller and that isn't used anywhere. But I'll leave it as a convenience, alongside `addStyleClassToLine()`/`removeStyleClassFromLine()`.

if it's not used id say it's fine to remove (just like us dropping `sql` support in this PR) 🎉

> Some manual checking done, but I still need to cover a few more cases.
> CodeMirror usage in Web Inspector is more prevalent than I expected 🤯

yeah that's why i suggested so much caution 😅

rcaliman-apple committed:
Web inspector fails to syntax highlight JS with backtick-delimited format strings
https://bugs.webkit.org/show_bug.cgi?id=283870
rdar://107619553

Reviewed by Devin Rousso and BJ Burg.

Update CodeMirror from version 5.21.1 to 5.65.20

Highlights
- Syntax highlighting for template literals (template strings)
- Styntax highlighting for private class elements

* Source/WebInspectorUI/Scripts/update-codemirror-resources.rb:
Check for the latest repository name format.
Stop copying over addon/mode/overlay.js and mode/sql/sql.js because they're not used anymore.

* Source/WebInspectorUI/Tools/CodeMirrorModes/index.html:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/LICENSE:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/clike.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/clojure.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/closebrackets.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/codemirror.css:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/codemirror.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/coffeescript.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/comment.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/css.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/htmlmixed.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/javascript.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/jsx.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/livescript.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/mark-selection.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/matchbrackets.js:

* Source/WebInspectorUI/UserInterface/External/CodeMirror/overlay.js: Removed.
`CodeMirror.overlayMode` Doesn't seem to be used anywhere.
This file has existed since Web Inspector inception and was periodically updated,
but whatever used this functionality was probably removed at some point.

* Source/WebInspectorUI/UserInterface/External/CodeMirror/placeholder.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/runmode.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/sass.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/searchcursor.js:

* Source/WebInspectorUI/UserInterface/External/CodeMirror/sql.js: Removed.
SQL syntax highlighting is no longer used.

* Source/WebInspectorUI/UserInterface/External/CodeMirror/sublime.js:
* Source/WebInspectorUI/UserInterface/External/CodeMirror/xml.js:
* Source/WebInspectorUI/UserInterface/Main.html:

* Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:
(WI.CSSStyleDeclaration.prototype.resolveVariableValue):
* Source/WebInspectorUI/UserInterface/Views/CodeMirrorAdditions.js:
* Source/WebInspectorUI/UserInterface/Views/CodeMirrorOverrides.css:
(.CodeMirror pre:is(.CodeMirror-line, .CodeMirror-line-like)):
(.CodeMirror pre): Deleted.
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetStyleProperty.js:
(WI.SpreadsheetStyleProperty.prototype._addGradientTokens):
(WI.SpreadsheetStyleProperty.prototype._addColorTokens):
(WI.SpreadsheetStyleProperty.prototype._addTimingFunctionTokens):
(WI.SpreadsheetStyleProperty.prototype._addVariableTokens):

In CodeMirror version 5.43.0, the token type for CSS functions was changed
from "atom" to "variable callee": https://github.com/codemirror/codemirror5/commit/3c79c54a335411fd1aff55246b77131dd98bb81b
Update references in Web Inspector which check against that token type.

* Source/WebInspectorUI/UserInterface/Views/SyntaxHighlightingDefaultTheme.css:
(.cm-s-default :is(.cm-tag, .cm-bracket, .cm-atom, .cm-variable.cm-callee, .cm-keyword, .cm-m-javascript.cm-builtin),):
(.cm-s-default :is(.cm-tag, .cm-bracket, .cm-atom, .cm-keyword, .cm-m-javascript.cm-builtin),): Deleted.

Update syntax highlighting theme to account for the new `cm-callee` class name set on CSS functions.

* Source/WebInspectorUI/UserInterface/Views/TextEditor.js:
(WI.TextEditor.prototype.toggleStyleClassForLine): Deleted.
toggleStyleClassForLine() is no longer used.

Canonical link: https://commits.webkit.org/300332@main
